File: sbt-bridge/src/xsbt/CompilerInterface.scala
/* sbt -- Simple Build Tool
 * Copyright 2008, 2009 Mark Harrah
 */
package xsbt

import xsbti.{ AnalysisCallback, Logger, Reporter, Severity }
import xsbti.compile._
import Log.debug
import java.io.File

import dotty.tools.dotc.core.Contexts.ContextBase
import dotty.tools.dotc.{ Main => DottyMain }
import dotty.tools.dotc.interfaces._

import java.net.URLClassLoader

final class CompilerInterface {
  def newCompiler(options: Array[String], output: Output, initialLog: xsbti.Logger,
                  initialDelegate: xsbti.Reporter): CachedCompiler = {
    // The classloader that sbt uses to load the compiler bridge is broken
    // (see CompilerClassLoader#fixBridgeLoader for details). To workaround
    // this we construct our own ClassLoader and then run the following code
    // with it:
    //   new CachedCompilerImpl(options, output)

    val bridgeLoader = getClass.getClassLoader
    val fixedLoader = CompilerClassLoader.fixBridgeLoader(bridgeLoader)
    val cciClass = fixedLoader.loadClass("xsbt.CachedCompilerImpl")
    cciClass.getConstructors.head
      .newInstance(options, output)
      .asInstanceOf[CachedCompiler]
  }

  def run(sources: Array[File], changes: DependencyChanges, callback: AnalysisCallback, log: Logger,
          delegate: Reporter, progress: CompileProgress, cached: CachedCompiler): Unit =
    cached.run(sources, changes, callback, log, delegate, progress)
}

class CachedCompilerImpl(args: Array[String], output: Output) extends CachedCompiler {
  val outputArgs =
    output match {
      case multi: MultipleOutput =>
        ???
      case single: SingleOutput =>
        List("-d", single.getOutputDirectory.getAbsolutePath.toString)
    }

  def commandArguments(sources: Array[File]): Array[String] =
    (outputArgs ++ args.toList ++ sources.map(_.getAbsolutePath).sortWith(_ < _)).toArray[String]

  def run(sources: Array[File], changes: DependencyChanges, callback: AnalysisCallback, log: Logger, delegate: Reporter, progress: CompileProgress): Unit = synchronized {
    run(sources.toList, changes, callback, log, delegate, progress)
  }
  private[this] def run(sources: List[File], changes: DependencyChanges, callback: AnalysisCallback, log: Logger, delegate: Reporter, compileProgress: CompileProgress): Unit = {
    debug(log, args.mkString("Calling Dotty compiler with arguments  (CompilerInterface):\n\t", "\n\t", ""))
    val ctx = (new ContextBase).initialCtx.fresh
      .setSbtCallback(callback)
      .setReporter(new DelegatingReporter(delegate))

    val cl = getClass.getClassLoader.asInstanceOf[URLClassLoader]

    val reporter = DottyMain.process(commandArguments(sources.toArray), ctx)
    if (reporter.hasErrors) {
      throw new InterfaceCompileFailed(args, Array())
    }
  }
}

class InterfaceCompileFailed(override val arguments: Array[String], override val problems: Array[xsbti.Problem]) extends xsbti.CompileFailed {
  override val toString = "Compilation failed"
}

File: sbt-bridge/src/xsbt/CompilerInterface.java
/* sbt -- Simple Build Tool
 * Copyright 2008, 2009 Mark Harrah
 */
package xsbt;

import xsbti.AnalysisCallback;
import xsbti.Logger;
import xsbti.Reporter;
import xsbti.compile.*;

import dotty.tools.dotc.Main;

import java.lang.reflect.InvocationTargetException;
import java.io.File;

/**
 * The new compiler interface is [[dotty.tools.xsbt.CompilerBridge]] that extends the new `xsbti.CompilerInterface2`.
 * This interface is kept for compatibility with Mill and the sbt 1.3.x series.
 */
public final class CompilerInterface {
  public CachedCompiler newCompiler(String[] options, Output output, Logger initialLog, Reporter initialDelegate) throws java.lang.Exception {
    if(isClassLoaderValid()) {
      return new CachedCompilerImpl(options, output);
    } else {
      initialLog.warn(() ->
        "The compiler class loader is badly configured.\n" +
        "Consider using a more recent version of your build tool:\n" +
        " - sbt >= 1.5.0\n" +
        " - Mill >= 0.9.3-21-002361\n" +
        " - Bloop >= 1.4.6-23-20a501bc"
      );
      // To workaround the wrong class loader, we construct our own and run
      // the following code with it:
      //   new CachedCompilerImpl(options, output)
      try {
        ClassLoader bridgeLoader = this.getClass().getClassLoader();
        ClassLoader fixedLoader = CompilerClassLoader.fixBridgeLoader(bridgeLoader);
        Class<?> cciClass = fixedLoader.loadClass("xsbt.CachedCompilerImpl");
        return (CachedCompiler) cciClass.getConstructors()[0].newInstance(options, output);
      } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
        throw new RuntimeException(e);
      }
    }
  }

  private boolean isClassLoaderValid() {
    // Check that the `xsbti.*` classes are loaded from the same class loader in the compiler and the bridge
    ClassLoader compilerClassLoader = Main.class.getClassLoader();
    Class<Logger> clazz = Logger.class;
    try {
      return compilerClassLoader.loadClass("xsbti.Logger") == clazz;
    } catch (ClassNotFoundException e) {
      throw new RuntimeException(e);
    }
  }

  public void run(File[] sources, DependencyChanges changes, AnalysisCallback callback, Logger log,
      Reporter delegate, CompileProgress progress, CachedCompiler cached) {
    cached.run(sources, changes, callback, log, delegate, progress);
  }
}
